backend/src/main/java/gov/cdc/usds/simplereport/service/TenantDataAccessService.java

```diff
@@ -0,0 +1,91 @@
+import java.util.List;
+import java.util.Optional;
+import java.util.Set;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.stereotype.Service;
+import org.springframework.transaction.annotation.Transactional;
+
+@Service
+@Transactional(readOnly = false)
+public class TenantDataAccessService {
+
+  private static final int VALID_MINUTES = 60;
+
+  private static final Logger LOG = LoggerFactory.getLogger(TenantDataAccessService.class);
+
+  @Autowired private TenantDataAccessRepository _repo;
+  @Autowired private CurrentTenantDataAccessContextHolder _contextHolder;
+  @Autowired private AuthorizationProperties _authProperties;
+  @Autowired private OrganizationExtractor _extractor;
+
+  public Optional<Set<String>> getTenantDataAccessAuthorityNames(ApiUser apiUser) {
+    List<TenantDataAccess> tenantDataAccessList =
+        _repo.findValidByApiUserId(apiUser.getInternalId());
+    if (tenantDataAccessList.isEmpty()) {
+      return Optional.empty();
+    } else if (tenantDataAccessList.size() != 1) {
+      // there should only be 1 valid tenant data access.  to clean this up, remove all
+      removeAllTenantDataAccess(apiUser);
+      return Optional.empty();
+    }
+
+    PermissionsData permissionsData = tenantDataAccessList.get(0).getPermissionsData();
+    return Optional.of(permissionsData.getAuthorities());
+  }
+
+  public Optional<OrganizationRoleClaims> addTenantDataAccess(
+      ApiUser apiUser, Organization org, String justification) {
+    // ensure user has no current valid tenant access
+    removeAllTenantDataAccess(apiUser);
+
+    Date expirationDate = Date.from(Instant.now().plus(Duration.ofMinutes(VALID_MINUTES)));
+
+    String prefix = _authProperties.getRolePrefix();
+    Set<String> authorities = new HashSet<>();
+
+    // retain site admin privileges, in the future it may be nice to drop these
+    authorities.add(_authProperties.getAdminGroupName());
+
+    // authority names for the org being accessed (assume org-level admin in the tenant)
+    authorities.add(prefix + org.getExternalId() + ":" + OrganizationRole.getDefault());
+    authorities.add(prefix + org.getExternalId() + ":" + OrganizationRole.ADMIN);
+
+    PermissionsData permissionsData = new PermissionsData(authorities);
+
+    _repo.save(new TenantDataAccess(apiUser, org, permissionsData, justification, expirationDate));
+
+    List<OrganizationRoleClaims> roleClaimsList =
+        _extractor.convertClaims(permissionsData.getAuthorities());
+    Optional<OrganizationRoleClaims> roleClaims = Optional.of(roleClaimsList.get(0));
+
+    LOG.info(
+        "** addTenantDataAccess for user: {} to org {}",
+        apiUser.getInternalId(),
+        org.getInternalId());
+
+    return roleClaims;
+  }
+
+  public void removeAllTenantDataAccess(ApiUser apiUser) {
+    _repo.findValidByApiUserId(apiUser.getInternalId()).forEach(i -> i.setIsDeleted(true));
```

Comment:
this query will end up leaving expired tenant data access records as un-deleted. is that what you want?

Reply:
That was the intent.  I was thinking that in the case that we needed to ever piece together an exact timeline of what happened that it seemed preferable to me to have expired records with `is_deleted` false and explicitly canceled records with `is_deleted` true with the `updated_at` time being the time of cancellation.  Maybe not?

Reply:
aren't these entities all audit-logged anyway?